[ONNX] Add support for tensordot operator.

fatcat-z opened the pull request:
Add the support for tensordot operator to exporter. It leverages einsum() operator to achieve the goal.
Also add 2 tests for this operator.

fatcat-z committed:
Now it supports shape with fixed size.

fatcat-z committed:
Now it supports shape with fixed size.

fatcat-z committed:
Now it supports shape with fixed size.

fatcat-z committed:
Now it supports shape with fixed size.

fatcat-z committed:
Remove loop for calculating sizes of reshape.

fatcat-z committed:
Remove loop for calculating sizes of reshape.

fatcat-z committed:
Remove loop for calculating sizes of reshape.

fatcat-z committed:
Finalize test and code of tensordot.

fatcat-z committed:
Finalize test and code of tensordot.

fatcat-z committed:
Finalize test and code of tensordot.

fatcat-z committed:
Merge branch 'new_tensordot' of https://github.com/fatcat-z/pytorch into new_tensordot

fatcat-z committed:
Resolve a lint issue.

fatcat-z committed:
Resolve lint issue.

jiafatom commented on torch/onnx/symbolic_opset12.py:
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
+
+    from torch.onnx.symbolic_opset9 import view, permute
+    for i in range(len(dims_a)):
+        if sym_help._get_tensor_dim_size(input_a, dims_a[i]) != sym_help._get_tensor_dim_size(input_b, dims_b[i]):
```
If both returns `None`, shall we also raise exception?

fatcat-z replied:
Good catch.

fatcat-z replied:
Has removed the check for the size here. Because sym_help._get_tensor_dim_size() will probably return a None in a reasonable scenario, like in my newly added test which uses dynamic_axes.

In that case, the size of those dimensions could not be got by calling sym_help._get_tensor_dim_size() method, but with my latest changes, we don't need to get the size values so None will has no impact to the function at all.

jiafatom commented on torch/onnx/symbolic_opset12.py:
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
+
+    dim_count_a = input_a.type().dim()
+    dim_count_b = input_b.type().dim()
+    for i in range(len(dims_a)):
```
Use list comprehension to write this?

jiafatom commented on torch/onnx/symbolic_opset12.py:
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
+    for i in range(len(dims_a)):
+        if (dims_a[i] < 0):
+            dims_a[i] += dim_count_a
+    for i in range(len(dims_b)):
```
list comprehension?

jiafatom commented on torch/onnx/symbolic_opset12.py:
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
+    new_input_b = permute(g, input_b, dims_b + left_dims_b)
+
+    output_a = view(g, new_input_a, left_dims_sizes_a + [-1])
+    output_a = view(g, output_a, [-1] + [sym_help._get_tensor_dim_size(output_a, len(left_dims_sizes_a))])
```
Can we directly use `view` once? Feel like the first `view` is only used to calculate the dim size.

fatcat-z replied:
In my latest commit, view() has been removed. Please help to take another review. Thanks

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -7831,6 +7831,32 @@ def forward(self, x, window_length: int):
+        x = torch.randint(6, (7, 5, 3, 4))
+        y = torch.randint(6, (3, 4, 8, 9))
+
+        self.run_test(M(), (x, y))
```
you can extend this test with dynamic input size (test with input of different sizes after export), by using argument `dynamic_axes` and `test_with_inputs`

fatcat-z replied:
Add one more test for this.

BowenBao commented on torch/onnx/symbolic_opset12.py:
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
+        if sym_help._get_tensor_dim_size(input_a, dims_a[i]) != sym_help._get_tensor_dim_size(input_b, dims_b[i]):
+            raise RuntimeError("Dimensions are not matched for operator tensordot: a - {}, b - {}.".format(dims_a, dims_b))
+
+    dim_count_a = input_a.type().dim()
```
I think there is helper function for getting rank in `sym_help` as well.

Also for this implementation, it would seem the tensor rank is required to be known as static number. This comes from limitation of potential `transpose/permute` later, which requires static rank, as well as from the code below for constructing dimensions. In that case, it would be better to throw not supported error when rank is unknown. (However, if `transpose` is not used, then we can lift the requirement on rank, by doing the below construction on dynamic shape with nodes of `ONNX::Shape`, and `ONNX::Slice`)

fatcat-z replied:
Update code to infer the size.

BowenBao commented on torch/onnx/symbolic_opset12.py:
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
+                   for i in range(dim_count_b)
+                   if (i not in dims_b)]
+
+    left_dims_sizes_a = [sym_help._get_tensor_dim_size(input_a, left_dims_a[i])
```
similar to David's comment, need to do None checks for returned values of `_get_tensor_dim_size`

BowenBao committed:
[ONNX] Fix assign input shape for tuple inputs & primitive type inputs (#54112)

shubhambhokare1 committed:
[ONNX] Update repeat_interleave symbolic (#54312)

Add implementation for cases when
- interleaving happens along dim which consist of dynamic axes

fatcat-z committed:
Update code for comments.

fatcat-z committed:
Update code for comments.

fatcat-z committed:
Remove trailing spaces.

fatcat-z committed:
Remove trailing spaces.

fatcat-z committed:
Fix lint comment.

fatcat-z committed:
Fix lint comment.

fatcat-z committed:
Resolve lint issue.

fatcat-z committed:
Resolve lint issue.

fatcat-z committed:
Remove trailing spaces after rebase.

fatcat-z committed:
Merge branch 'new_tensordot' of https://github.com/fatcat-z/pytorch into new_tensordot

jiafatom commented on torch/csrc/jit/passes/erase_number_types.cpp:
```diff
@@ -6,13 +6,19 @@
-      if (inp->type()->isSubtypeOf(NumberType::get())) {
-        inp->setType(TensorType::get());
-      }
+      SetNumTypeToTensorType(inp);
```
We originally only deal with NumberType not BoolType, now it considers BoolType as well. Is this on purpose?

jiafatom commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -67,6 +67,8 @@ def _shape_as_tensor(g, input):
 
 
 def _reshape_from_tensor(g, input, shape):
+    if (type(shape) is list):
```
isinstance(shape, list)?
